src/platform/intel/ace/include/ace/lib/mailbox.h

```diff
@@ -39,29 +39,29 @@
 static inline void mailbox_sw_reg_write(size_t offset, uint32_t src)
 {
 	volatile uint32_t *ptr;
-	volatile uint32_t __sparse_cache *ptr_c;
+	uint32_t __sparse_cache *ptr_c;
 
-	ptr_c = (volatile uint32_t __sparse_cache *)(MAILBOX_SW_REG_BASE + offset);
+	ptr_c = (uint32_t __sparse_cache *)(MAILBOX_SW_REG_BASE + offset);
```

Comment:
This looks a bit weird to me (we are casting volatile back), but compiler seems to be happy, so I guess this is ok. You live and learn.

Reply:
well, first of all you know "volatile is evil." But the point here is that that pointer is never used to actually access memory, so no need for `volatile` there at all

Reply:
> "volatile is evil."

I think "harmful" is more popular hence a better search keyword: https://www.bing.com/search?q=volatile+considered+harmful

Reply:
> if I am sharing memory between devices (to a context where the compiler doesn't see that the memory will be accessed) then volatile is appropriate.

@paulstelian97 not sure which "devices" you mean. I think `volatile` is supposed to tell the compiler that contents of that memory can change or have effects outside of the scope of this code, so don't optimise access out.
So e.g.
```
*v = 1;
*v = 2;
```
would normally look redundant to the compiler and it might throw the first assignment away, but if the `v` pointer is declared as `volatile`, it won't. Similar for
```
X = *v;
Y = *v;
```
the compiler could reuse the value already read from `v` in the first line to effectively make the second like `Y = X;` but if `v` is a `volatile` pointer, it won't.
In this case that pointer points to memory, shared with the host, so the compiler has no way to know that the host can modify that memory. But to the compiler `ptr` and `ptr_c` are two different pointers, pointing to two different addresses - with some fixed offset. The compiler has no idea that it's "the same" memory. So whatever we do with one of the two pointers will not tell anything to the compiler that something also affects the other pointer. And since `ptr_c` is never used to access memory, since it is never dereferenced, the `volatile` qualifier for it has no effect.